src/contracts/middleware/OBaseMiddlewareReader.sol

```diff
@@ -780,6 +796,121 @@ contract OBaseMiddlewareReader is
         return (false, hex"");
     }
 
+    function _getOperatorPowerPerVault(
+        uint256 maxNumOperatorsToCheck,
+        uint256 cacheIndex,
+        uint48 timestamp,
+        PauseableEnumerableSet.AddressSet storage operators,
+        uint256 operatorsLength_
+    )
+        private
+        view
+        returns (IMiddleware.VaultPower[] memory vaultsPower, bool atLeastOneActive, uint256 operatorsProcessed)
+    {
+        VaultManagerStorage storage $ = _getVaultManagerStorage();
+        address[] memory sharedVaults = $._sharedVaults.getActive(timestamp);
+
+        vaultsPower = new IMiddleware.VaultPower[](sharedVaults.length);
+
+        IMiddleware.AccumulateCtx memory ctx;
+        ctx.timestamp = timestamp;
+        ctx.subnetwork = _NETWORK().subnetwork(0).identifier();
+        uint256 i;
+
+        for (i = cacheIndex; i < cacheIndex + maxNumOperatorsToCheck && i < operatorsLength_;) {
+            (ctx.totalVaultsIndex, atLeastOneActive) = _accumulateOperatorVaultPower(vaultsPower, operators, i, ctx);
+            unchecked {
+                ++i;
+            }
+        }
+
+        operatorsProcessed = i - cacheIndex;
+
+        uint256 totalVaultsIndex = ctx.totalVaultsIndex;
+
+        // Shrink array length to actual number of used slots
+        assembly ("memory-safe") {
+            mstore(vaultsPower, totalVaultsIndex)
+        }
+    }
+
+    function _accumulateOperatorVaultPower(
+        IMiddleware.VaultPower[] memory vaultsPower,
+        PauseableEnumerableSet.AddressSet storage operators,
+        uint256 index,
+        IMiddleware.AccumulateCtx memory ctx
+    ) private view returns (uint256, bool) {
+        (address operator, uint48 enabled, uint48 disabled) = operators.at(index);
+
+        if (enabled < ctx.timestamp && (disabled == 0 || disabled >= ctx.timestamp)) {
+            ctx.atLeastOneActive = true;
+
+            address[] memory operatorVaults = _activeVaultsAt(ctx.timestamp, operator);
+
+            for (uint256 j; j < operatorVaults.length;) {
+                address vault = operatorVaults[j];
+                uint256 power = _getOperatorPowerAt(ctx.timestamp, operator, vault, ctx.subnetwork);
+
+                if (power != 0) {
+                    ctx.totalVaultsIndex += _extractVaultsPower(vaultsPower, ctx.totalVaultsIndex, vault, power);
+                }
+
+                unchecked {
+                    ++j;
+                }
+            }
+        }
+
+        return (ctx.totalVaultsIndex, ctx.atLeastOneActive);
+    }
+
+    function _extractVaultsPower(
+        IMiddleware.VaultPower[] memory vaultsPower,
+        uint256 totalVaultsIndex_,
+        address vault,
+        uint256 power
+    ) private pure returns (uint256 totalVaultsIndex) {
+        // Inner accumulation loop
+        for (uint256 k; k < vaultsPower.length; ++k) {
```

Comment:
I'm guessing this was extracted and `k` used to be needed, but can we use `i` instead of `k` now?